### PR TITLE
Quickcache UserRole.get and memoize User.role

### PR DESCRIPTION
**Pull request by esoergel**
@gcapalbo @dannyroberts @czue 
`UserRole.get` is called only for non-admin users, so we devs didn't really
feel the pain of that.  On the dashboard page, I saw it get called 68 times
for one page load, and for the base reports page, 144 times.

Turns out memoizing `User.role` was really all that was necessary.  I figure
slapping a `QuickCacheDocumentMixin` in here is super easy too and doesn't
hurt anything, so I did that too (it'll span multiple requests, and I
really can't imagine roles changing much).

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -336,11 +336,15 @@ def get_default(cls, domain=None):
...
     @classmethod
     def role_choices(cls, domain):
-        return [(role.get_qualified_id(), role.name or '(No Name)') for role in [AdminUserRole(domain=domain)] + list(cls.by_domain(domain))]
+        return [(role.get_qualified_id(), role.name or '(No Name)') for role in
+                [AdminUserRole(domain=domain)] + list(cls.by_domain(domain))]
 
     @classmethod
     def commcareuser_role_choices(cls, domain):
-        return [('none','(none)')] + [(role.get_qualified_id(), role.name or '(No Name)') for role in list(cls.by_domain(domain))]
+        return [('none','(none)')] + [
```
Comment (dimagimon) [inline corehq/apps/users/models.py]: E231 missing whitespace after ','

**Comment by czue** [comment]
loving the return of operation :cat2: !

File: corehq/util/quickcache.py [hunk 2]
```diff
@@ -290,7 +290,7 @@ def domains_for_user(request):
             CacheWithTimeout(
                 django_caches['locmem'],
                 timeout=memoize_timeout)]
-        if timeout != 0:
+        if timeout > 0:
```
Comment (gcapalbo) [inline corehq/util/quickcache.py]: Can `timeout` ever be `None`? If so this would product different behavior (not sure if that matters or not)

**Comment by gcapalbo** [comment]
(looks like the build failure is related to this change as well)

**Comment by esoergel** [comment]
Yeah, build failure looks like it's probably caused by memoizing `User.role`.  I'll investigate.

File: corehq/util/quickcache.py (hunk 2 above)
Comment (esoergel) [inline corehq/util/quickcache.py]: no, not as the code currently stands.  Line 285 above:
```python
timeout = timeout if timeout is not None else 5 * 60
```
This means the only way to never timeout is to use something like `timeout=BIGNUM`, which is probably what we want.

**Comment by dannyroberts** [comment]
Looks good!

**Pull request merged**